File: Exercise_1.py
# ## Problem1
# Custom Sort String (https://leetcode.com/problems/custom-sort-string/)
# // Time Complexity :O(n)
# // Space Complexity :O(n)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No


# // Your code here along with comments explaining your approach
class Solution:
    def customSortString(self, order, s):
        # creating a hashmap of chars in s
        hmap_s = {}
        for i in range(len(s)):
            hmap_s[s[i]] = hmap_s.get(s[i], 0) + 1
        result = ""
        for char in order:
            if char in hmap_s:
                result += char * hmap_s[char]
                del hmap_s[char]
        for char, count in hmap_s.items():
            result += char * count
        return result


order = "cba"
s = "abcd"
sol = Solution()
print(sol.customSortString(order, s))

File: Exercise_1.cpp
/* 
    Time Complexity                              :  O(M+N) - where M  is the size of the order string and
                                                            N is the size of the s string.
    Space Complexity                             :  O(N) - to store the characters from s string
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
public:
    string customSortString(string order, string s) {
        
        return hashMapApproach(order, s);
    }
    
    string hashMapApproach(string order, string s) {
        unordered_map<char, int> mp;
        for(auto ch : s) {
            mp[ch]++;
        }
        string ans = "";
        for(auto ch : order) {
            if(mp.find(ch) != mp.end()) {
                for(int i=0;i<mp[ch];i++) {
                    ans += ch;
                }
                mp.erase(ch);
            }
        }
        
        for(auto m : mp) {
            for(int i=0;i<m.second;i++) {
                ans+=m.first;
            }
        }
        
        return ans;
    }
};